Add formula for gn

ry opened the pull request:
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This is needed to build V8 and Deno (#35645) or any other Chromium project.

SMillerDev commented on Formula/gn.rb:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/gn --version")
```
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

ry replied:
@SMillerDev PTAL, I have added a more robust test

hayd commented on Formula/gn.rb:
```diff
@@ -0,0 +1,44 @@
+    system "python", "build/gen.py"
+    system "ninja", "-C", "out/", "gn"
+    mkdir bin
+    cp "out/gn", bin/"gn"
```
nit: I think these two lines are equivalent to `bin.install "out/gn"`

SMillerDev replied:
They are, @ryland could you edit this?

ry replied:
Done

hayd commented:
@SMillerDev is the build error something independent of this PR? https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/40433/version=mojave/testReport/brew-test-bot/mojave/doctor/

ry commented:
Rebased and squashed. I think it's an unrelated error.

hayd commented:
Yep, looks like all PRs are failing with that same issue.

ry commented:
Depends on https://github.com/Homebrew/homebrew-core/issues/38717

ry commented:
@SMillerDev @hayd I think this is ready. PTAL.

fxcoudert commented:
@SMillerDev I'm not sure this has a stable release system, it looks like the formula is just pulling a “random” revision from head.

ry commented:
gn's master branch is the release branch. https://gn.googlesource.com/gn

fxcoudert commented:
From https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae: 

> Formulae in the core repository must have a stable version tagged by the upstream project. […]
> We don’t accept software without a tagged version because they regularly break due to upstream changes and we can’t provide bottles for them.

ry commented:
@fxcoudert gn doesn't have tagged releases. Note that currently homebrew's v8 crate downloads a binary of gn (via depot_tools) - which also is against the rules - this crate would fix that. **All of chromium** depends on gn's release branch (master) - it's not a random github project from an unknown user. The build will not break on upgrade - I think an exception is warranted. Otherwise please suggest a course of action.

SMillerDev commented:
I’d say pulling in random dependencies is worse then no tagged release. @ry where did you get the    `version "1547"` from though? Because that might be good to document if we're going that direction.

MikeMcQuaid commented:
> Note that currently homebrew's v8 formula downloads a binary of gn (via depot_tools) - which also is against the rules - this formula would fix that.

It could be part of the `v8` formula instead in which relying on a specific, known-to-work commit would be the way forward. We cannot include this without a stable release, sorry.

hayd commented:
@mikemcquaid

> Otherwise please suggest a course of action.

This is required for deno and seemingly required for v8 too (if it wasn't to use a fetched binary). What is the way forward here?

ry commented:
GN does not tag versions, but Chromium does and specifies a GN commit id. 

Here is what Chrome version 75.0.3767.2 uses:
 https://github.com/chromium/chromium/blob/75.0.3767.2/DEPS#L291

I've updated the formula to use this id and added some documentation so that when this is upgraded the maintainer knows where to look for a version that builds Chromium.

@MikeMcQuaid @SMillerDev PTAL.

MikeMcQuaid commented:
> I've updated the formula to use this id and added some documentation so that when this is upgraded the maintainer knows where to look for a version that has built a Chromium release.

@ry Perfect, thanks.

MikeMcQuaid commented:
@ry Actually, sorry, misread that: I think that that should instead be used for a `resource` in the `v8` and `deno` formulae instead.

ry commented:
@MikeMcQuaid gn is a useful standalone build system - much like cmake or bazel. I think users would benefit from having it in homebrew. 
(I would argue that it’s the best open source build system - but suffers because it’s not available in distributions like homebrew.)

MikeMcQuaid commented:
@ry I don't disagree but we need it to have stable, tagged releases to be included in Homebrew.

ry commented:
@MikeMcQuaid It is a stable tagged release - it’s just not tagged with git tags.

I can fork Google Chrome’s source code into my own personal Github and then git tag it? I don’t think that’s very secure tho.

Please suggest a course of action.

MikeMcQuaid commented:
> Please suggest a course of action.

I have. It's not going to get included without Git tags. This is documented: https://docs.brew.sh/Acceptable-Formulae#stable-versions

> I’m not sure what you’re looking for.

For any project that uses/needs this (e.g. `v8` and `deno`) to use a `resource` block in the formula instead of a separate formula being created.

ry commented:
@MikeMcQuaid

> Formulae in the core repository must have a stable version tagged by the upstream project. 

This is satisfied: https://github.com/chromium/chromium/blob/75.0.3767.2/DEPS#L290-L291

MikeMcQuaid commented:
@ry That is not a tag in the upstream `gn` project. There is not a tag in the upstream project. A "tag" has a specific meaning in the context of a Git repository: that is not it. Arguing with me is a waste of both of our time. Locking.
